### PR TITLE
[Easy] make copy constructed tensor a leaf variable when using torch.tensor(sourceTensor)

### LINKED ISSUE #10876: [pytorch] requires_grad=False from torch.tensor ignored when the input tensor has requires_grad=True
## Issue description

As shown by the following example, the expected behavior (at least from me) is to copy only the data from `t1` to `t2`, however `torch.tensor()` seems to build up a computation graph from `t1` instead ...

## Code example

```python
>>> t1 = torch.tensor([1,1,1], dtype=torch.float, requires_grad=True)
>>> t2 = torch.tensor(t1, requires_grad=False)
>>> t2.requires_grad
True
>>> t1
tensor([1., 1., 1.], requires_grad=True)
>>> t2
tensor([1., 1., 1.], grad_fn=<CopyBackwards>)
```

## System Info

- PyTorch or Caffe2: PyTorch
- How you installed PyTorch (conda, pip, source): conda
- OS: Ubuntu 16.04
- PyTorch version: 0.4.1
- Python version: 3.6.3

### PR DESCRIPTION
**Pull request by weiyangfb**
- fix https://github.com/pytorch/pytorch/issues/10876
- the cause of the bug is because copy constructor cannot distinguish between default value of requires_grad and requires_grad=False, thus it makes a copy from source tensor along with its grad_fn if requires_grad=True at source
- with this fix, the behavior becomes
```
>>> source = torch.randn(2, 2, requires_grad=True)
>>> copy = torch.tensor(source, requires_grad=True)
>>> print(copy)
tensor([[-1.2001,  1.9869],
        [-1.0134,  1.3096]], grad_fn=<CopyBackwards>)

>>> source = torch.randn(2, 2, requires_grad=True)
>>> copy = torch.tensor(source, requires_grad=False)
>>> print(copy)
tensor([[-0.7402,  0.0467],
        [ 0.4344, -0.0420]])

>>> source = torch.randn(2, 2, requires_grad=True)
>>> copy = torch.tensor(source)
>>> print(copy)
tensor([[-0.7402,  0.0467],
        [ 0.4344, -0.0420]])
```

### CODE REVIEW

**Comment by weiyangfb** [comment]
@ezyang can I get a review on this easy PR?

File: torch/csrc/utils/tensor_new.cpp [hunk 1]
```diff
@@ -211,10 +212,15 @@ Tensor internal_new_from_data(const Type & type, at::optional<Device> device_opt
                                                        *torch::getLayout(type.backend()),
                                                        type_inference_device_type);
       const auto& type_to_use = type_inference ? type_inference_type : type;
-      return copy_variables ? new_with_tensor_copy(type_to_use, var, device_index) :
+      // temporarily mute requires_grad when copy from var with requires_grad = false
```
Comment (ezyang) [inline torch/csrc/utils/tensor_new.cpp]: Eeeeew. I don't like it. What if a concurrent thread is reading from this variable? Surely there's a way to do this without mutating.

**Review by ezyang — CHANGES_REQUESTED** [review]
Read-only ops should not mutate.

**Comment by apaszke** [comment]
I would also say that the semantics are slightly weird. The output variable depends on the data of the input variable, and so there's no reason not to differentiate through this...

File: torch/csrc/utils/tensor_new.cpp [hunk 2]
```diff
@@ -211,10 +212,15 @@ Tensor internal_new_from_data(const Type & type, at::optional<Device> device_opt
...
+      if (mute_grad) var.set_requires_grad(false);
+      auto new_tensor = copy_variables ? new_with_tensor_copy(type_to_use, var, device_index) :
                               new_with_type_conversion(type_to_use, var, device_index);
+      if (mute_grad) var.set_requires_grad(true);
+      return new_tensor;
```
Comment (zou3519) [inline torch/csrc/utils/tensor_new.cpp]: Can't you set requires_grad = true on the new_tensor? that would avoid concurrency issues because no other threads hav access to the new tensor yet

Comment (weiyangfb) [reply torch/csrc/utils/tensor_new.cpp]: I guess you meant to set `requires_grad=false` on the new_tensor. Yes, I can do that if it is sufficient. Please note that `grad_fn` will still be created for new_tensor at [s_copy_](https://github.com/pytorch/pytorch/blob/9a7c196040f55d921a4deee42cbb1dd294b8f23b/tools/autograd/templates/VariableType.cpp#L408-L415). If we need to be more careful, then I can also pass in a `requires_grad` args to `s_copy_` to avoid silently ignored `requires_grad` at `torch.tensor()`. Which way is preferred? @ezyang @apaszke

**Comment by weiyangfb** [comment]
@apaszke I would guess the copy constructed tensor should not be made differentiable through `copy` op here. Please correct me if I am wrong, I think the expected behavior should be the same as:
```
source = torch.randn(10, requires_grad=True)
copy = torch.tensor(source.data)
```

**Comment by ezyang** [comment]
To figure out a good solution to this problem, we must ask ourselves:

1. How is `tensor()` with a `Tensor` argument implemented today, and
2. Why does this implementation "do the wrong thing" with respect to `requires_grad`?

We can see that `tensor()` dispatches to `tensor_ctor`, which dispatches to `internal_new_from_data`. Here is the important lines in the function:

```
  if (THPVariable_Check(data)) {
      auto var = reinterpret_cast<THPVariable*>(data)->cdata;
      auto type_inference_device_type = device_opt.has_value() ? device_opt->type()
                                                               : torch::getDeviceType(var.type());
      // infer the scalar type and device type; it's not expected to infer the layout since these constructors
      // are defined per-layout-type (e.g. tensor vs sparse_coo_tensor).
      const auto& type_inference_type = torch::getVariableType(var.type().scalarType(),
                                                       *torch::getLayout(type.backend()),
                                                       type_inference_device_type);
      const auto& type_to_use = type_inference ? type_inference_type : type;
      return copy_variables ? new_with_tensor_copy(type_to_use, var, device_index) :
                              new_with_type_conversion(type_to_use, var, device_index);
  }
```

In the bodies of `new_with_tensor_copy` and  `new_with_type_conversion` you can see this calls `copy` or `toType` on **Variable**, which means you get the normal gradient behavior as if you had called `copy()` or `toType()` on them, which is why you incorrectly get `requires_grad` at the end.

Based on this, this suggests two solutions:

1. Keep calling `copy`/`toType`, but then `detach` the tensor afterwards, so that you set `requires_grad = false` and delete the `grad_fn`. Effectively, you are saying that `tensor(t)` is equivalent to `t.clone().detach()`. This is the "do the wrong thing first, and then fixup after yourself" strategy.
2. Create a new operator on VariableType, `new_with_tensor`, which does the right thing from the get go and doesn't actually ever create a grad fn or set requires grad on the output. This is the "do the right thing from the beginning" strategy, but it is more complicated to implement.

I think both approaches are reasonable. If you want to do (2), you'll need to write a custom `VariableType` method implementation, based on the fact that detach is hardcoded, like this:

```
Tensor VariableType::detach(const Tensor & self) const {
  profiler::RecordFunction profiler("detach");
  torch::jit::Node* node = nullptr;
  if (jit::tracer::isTracing()) {
    auto& graph = jit::tracer::getTracingState()->graph;
    node = graph->create(jit::aten::detach, /*outputs=*/0);
    jit::tracer::recordSourceLocation(node);
    jit::tracer::addInputs(node, "self", self);
    graph->appendNode(node);

  }
  // <NON_GENERATED_CODE>
  auto result = as_variable_ref(const_cast<Tensor&>(self)).detach();
  // </NON_GENERATED_CODE>
  if (jit::tracer::isTracing()) {
    jit::tracer::addOutput(node, result);
  }
  return result;
}
```

You'll have something similar, except that it calls `copy`/`toType` as necessary. But I think detach after the fact is also a very reasonable strategy, and much simpler.

**Comment by weiyangfb** [comment]
@ezyang I am picking the (2) approach here: `clone()` and `detach()` when copy construct a new tensor.

File: torch/csrc/utils/tensor_new.cpp [hunk 3]
```diff
@@ -463,13 +469,17 @@ Tensor tensor_ctor(const Type& type, PyObject* args, PyObject* kwargs) {
   auto r = parser.parse(args, kwargs, parsed_args);
   if (r.idx == 0) {
     bool type_inference = r.isNone(1);
+    // args_requires_grad=true if requires_grad is not set or requires_grad=True
+    // args_requires_grad=false requires_grad is set and requires_grad=false
```
Comment (ezyang) [inline torch/csrc/utils/tensor_new.cpp]: err, when you say `requires_grad=false`, you mean that it's set, right? So it would be more accurate to just say "args_requires_grad=false  if explicitly requires_grad=False"

**Review by ezyang — CHANGES_REQUESTED** [review]
I feel there are some semantic problems.

Basically, in my opinion (with concurrences from @gchanan  and @apaszke), we think that `torch.tensor(x)` should be equivalent to `x.clone().detach()` and `torch.tensor(x, requires_grad=True)` should be equivalent to `x.clone().detach().requires_grad_(True)`. The basic idea is that `torch.tensor` reads out "the data" from whatever it is passed, and constructs a leaf variable. This gives us the invariant that "for any argument x, `torch.tensor(x)` does not require grad, and `torch.tensor(x, requires_grad=True)` gives a leaf variable that requires grad).

**Comment by weiyangfb** [comment]
@ezyang Thanks a lot for the clarifications! Sorry I missed the default case. I will make changes so that:
```
source = torch.randn(3, 3, requires_grad=True)
copy = torch.tensor(source) # requires_grad=False
```

File: torch/csrc/utils/tensor_new.cpp [hunk 4]
```diff
@@ -463,13 +465,16 @@ Tensor tensor_ctor(const Type& type, PyObject* args, PyObject* kwargs) {
   auto r = parser.parse(args, kwargs, parsed_args);
   if (r.idx == 0) {
     bool type_inference = r.isNone(1);
+    // args_requires_grad=True only if requires_grad is set and is set to True (requires_grad=True)
+    bool args_requires_grad = !r.isNone(3) && r.toBool(3);
```
Comment (ezyang) [inline torch/csrc/utils/tensor_new.cpp]: I'm surprised that with the default specification `bool requires_grad=False` you need to test the `isNone` case...

File: torch/csrc/utils/tensor_new.cpp [hunk 5]
```diff
@@ -216,8 +216,10 @@ Tensor internal_new_from_data(const Type & type, at::optional<Device> device_opt
                                                        *torch::getLayout(type.backend()),
                                                        type_inference_device_type);
       const auto& type_to_use = type_inference ? type_inference_type : type;
-      return copy_variables ? new_with_tensor_copy(type_to_use, var, device_index) :
+      auto new_tensor = copy_variables ? new_with_tensor_copy(type_to_use, var, device_index) :
                               new_with_type_conversion(type_to_use, var, device_index);
+      if (!args_requires_grad) new_tensor.detach_();
```
Comment (ezyang) [inline torch/csrc/utils/tensor_new.cpp]: No, this still looks wrong. Based on the proposed semantics in my comment, you should *unconditionally* detach, and then depending on `args_require_grad` set the `requires_grad` back to True (turning it into a requires_grad leaf variable).

File: test/test_torch.py [hunk 6]
```diff
@@ -2364,6 +2364,28 @@ def test_tensor_factory(self):
...
+        self.assertEqual(copy.data, source.data)
+        self.assertTrue(source.requires_grad)
+        self.assertFalse(copy.requires_grad)
+
+        # copy with requires_grad=True
+        source = torch.randn(5, 5, requires_grad=True)
+        copy = torch.tensor(source, requires_grad=True)
+        self.assertEqual(copy.data, source.data)
+        self.assertTrue(source.requires_grad)
+        self.assertTrue(copy.requires_grad)
```
Comment (ezyang) [inline test/test_torch.py]: These tests should test leaf status.

**Review by ezyang — CHANGES_REQUESTED** [review]
I think it's still not right.

Also, can we update the docs for `tensor` for this case?

**Comment by ezyang** [comment]
In general, we should discourage people from passing tensor to `torch.tensor`, in favor of explicitly saying `x.clone().detach()` or `x.clone().detach().requires_grad_(True)`. @apaszke suggested a warning might be appropriate in this case.

File: torch/_torch_docs.py [hunk 7]
```diff
@@ -3818,6 +3818,13 @@ def parse_kwargs(desc):
     If you have a NumPy ``ndarray`` and want to avoid a copy, use
     :func:`torch.from_numpy`.
 
+.. warning::
+
+    When data is a tensor `x`, :func:`torch.tensor` reads out 'the data' from whatever it is passed,
+    and constructs a leaf variable. Therefore ``torch.tensor(x)`` is equivalent to ``x.clone().detach()``
+    and ``torch.tensor(x, requires_grad=True)`` is equivalent to ``x.clone().detach().requires_grad_(True)``.
+    The equivalences use ``clone()`` and ``detach()`` are recommended.
```
Comment (ezyang) [inline torch/_torch_docs.py]: nit: *equivalent?

**Review by ezyang — APPROVED** [review]

File: torch/csrc/utils/tensor_new.cpp [hunk 8]
```diff
@@ -216,8 +220,11 @@ Tensor internal_new_from_data(const Type & type, at::optional<Device> device_opt
                                                        *torch::getLayout(type.backend()),
                                                        type_inference_device_type);
       const auto& type_to_use = type_inference ? type_inference_type : type;
-      return copy_variables ? new_with_tensor_copy(type_to_use, var, device_index) :
+      auto new_tensor = copy_variables ? new_with_tensor_copy(type_to_use, var, device_index) :
                               new_with_type_conversion(type_to_use, var, device_index);
+      new_tensor.detach_(); // making copy constructed tensor a leaf node
```
Comment (apaszke) [inline torch/csrc/utils/tensor_new.cpp]: I'm pretty sure the `new_with_type_conversion` branch doesn't guarantee that it won't return the exact same variable (this is why we have the `copy_variables` flag!!), so this code is incorrect, and can modify the input to e.g. `torch.as_tensor`.

**Review by apaszke — COMMENTED** [review]
I'm not sure if what we did here is correct.

File: torch/csrc/utils/tensor_new.cpp [hunk 9]
```diff
@@ -207,6 +207,10 @@ Tensor internal_new_from_data(const Type & type, at::optional<Device> device_opt
   }
 
   if (THPVariable_Check(data)) {
+      PyErr_WarnEx(PyExc_UserWarning,
+        "To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() "
```
Comment (gchanan) [inline torch/csrc/utils/tensor_new.cpp]: you should mention `torch.as_tensor` here for the use case when you just want to get a tensor out (for data that may be mixed between e.g. tensors, numpy arrays, lists), and don't want to copy the data.

File: torch/csrc/utils/tensor_new.cpp [hunk 10]
```diff
@@ -463,13 +470,16 @@ Tensor tensor_ctor(const Type& type, PyObject* args, PyObject* kwargs) {
...
     return internal_new_from_data(
                typeWithDefault(r, 1, 2, type),
                r.deviceOptional(2),
                r.pyobject(0),
                true,
                true,
-               type_inference)
+               type_inference,
+               args_requires_grad)
         .set_requires_grad(r.toBool(3));
```
Comment (gchanan) [inline torch/csrc/utils/tensor_new.cpp]: nit: set this to args_requires_grad so you don't have to read out the arguments again.

File: torch/csrc/utils/tensor_new.cpp (hunk 8 above)
Comment (gchanan) [reply torch/csrc/utils/tensor_new.cpp]: Yes, this is incorrect.  I don't think passing `args_requires_grad` is the right mechanism here: it's not even checked in most of the branches, but there's no indication that this should be the case.  Just move the detach / set logic to where you need it (tensor_new and tensor_ctor).

Comment (weiyangfb) [reply torch/csrc/utils/tensor_new.cpp]: Thanks for pointing this out! Sorry I landed it without thinking it thoroughly. I will push a new PR to fix this

File: torch/csrc/utils/tensor_new.cpp (hunk 9 above)
Comment (weiyangfb) [reply torch/csrc/utils/tensor_new.cpp]: actually I might just put warnings at `tensor_ctor` and `new_tensor` instead.
